src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -1866,6 +1872,20 @@ fn specialize<'p, 'a: 'p, 'tcx>(
                     "unexpected ctor {:?} for slice pat", constructor)
             }
         }
+
+        PatternKind::Or { ref pats } => {
+            let mut specialized_pats = vec![];
+            for pat in pats {
+                if let Some(s) = specialize(cx, &[pat], constructor, wild_patterns) {
+                    specialized_pats.extend(s);
```

Comment:
This is flattening the arrays, but I think there ought to be a separate vector for each pattern. See:
https://github.com/rust-lang/rust/blob/4c66ed3ce0091b8ba5c838bd4e650adc6887d7ff/src/librustc_mir/hair/pattern/_match.rs#L61-L63

Reply:
I think you're right. I removed the specialization code, but in the next PR I'll check into this.